554 Flow API for one-time and periodic tasks

maimoonak opened the pull request:
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #554 

**Description**
Implementation of FlowAPI for managing sync states and last sync date for one-time and periodic sync

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR? WIP

**Type**
Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

maimoonak commented:
@f-odhiambo Here is the PR for feedback on last sync date issue

kevinmost commented on engine/src/main/java/com/google/android/fhir/SharedPreferencesUtil.kt:
```diff
@@ -0,0 +1,68 @@
+    val prefsEditor: SharedPreferences.Editor = prefs.edit()
+    with(prefsEditor) {
+      putString(key, value)
+      commit()
```
`commit` blocks this thread. If you use `apply`, it wouldn't block, but then you have to deal with possibly-stale values.

Maybe we can make this a `suspend fun write` and keep `commit`, so that it's essentially asynchronous? @yigit any opinions?

yigit replied:
We should use jetpack data store instead of shared preferences. 
Even though Shared Prefs have an async write API, it might still block the UI thread if a write is pending while context is being destroyed.

kevinmost commented on engine/src/main/java/com/google/android/fhir/SharedPreferencesUtil.kt:
```diff
@@ -0,0 +1,68 @@
+    val prefsEditor: SharedPreferences.Editor = prefs.edit()
+    with(prefsEditor) {
+      putLong(key, value)
+      commit()
```
Same question here for @yigit

yigit replied:
Sorry I didn't read the whole CL (ooo) but unless I'm missing something, we should use data store,not shred preferences.
Data store also provides first class coroutines API, which is what we want for internals

kevinmost commented on engine/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt:
```diff
@@ -27,12 +28,15 @@ import com.google.android.fhir.search.Search
 /** Implementation of [FhirEngine]. */
-internal class FhirEngineImpl
-constructor(private val database: Database, private val context: Context) : FhirEngine {
+internal class FhirEngineImpl constructor(private val database: Database, context: Context) :
```
Nit: The `constructor` keyword is unnecessary here, remove it.

kevinmost commented on engine/src/main/java/com/google/android/fhir/sync/FhirSynchronizer.kt:
```diff
@@ -31,6 +33,20 @@ sealed class Result {
+  object Started : State()
+
+  data class InProgress(val resourceType: ResourceType?) : State()
+  data class Glitch(val exceptions: List<ResourceSyncException>) : State()
```
What is the difference between `Glitch` and `Error`?

maimoonak replied:
Error or Success is final state emitted when Synchronizer finishes and sends final result. The Glitch is the intermediate error it faces and still continues. A Glitch may lead to Success or Error at the end, but we are enabling consumer to read intermediates states and progress for all resource types

kevinmost replied:
Makes sense. Please add docs on each of these state classes explaining the difference.

kevinmost commented on engine/src/main/java/com/google/android/fhir/sync/SyncJob.kt:
```diff
@@ -0,0 +1,34 @@
+    periodicSyncConfiguration: PeriodicSyncConfiguration,
+    context: Context,
+    clazz: Class<W>
+  ): Flow<MutableList<WorkInfo>>
```
Change this to `Flow<List<WorkInfo>>`. We shouldn't pass around mutable collections.

kevinmost commented on engine/src/main/java/com/google/android/fhir/sync/SyncJob.kt:
```diff
@@ -0,0 +1,34 @@
+  ): Flow<MutableList<WorkInfo>>
+
+  suspend fun run(subscribeTo: MutableSharedFlow<State>?): Result
+  fun workInfoFlowFor(uniqueWorkerName: String, context: Context): Flow<WorkInfo>
+  fun stateFlowFor(uniqueWorkerName: String, context: Context): Flow<State>
```
Let's change `uniqueWorkerName: String` to `worker: SyncWorkType`, and let the implementation call `worker.workerName` internally.

kevinmost commented on engine/src/main/java/com/google/android/fhir/sync/SyncJobImpl.kt:
```diff
@@ -0,0 +1,118 @@
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.mapNotNull
+
+class SyncJobImpl(
```
All of the public-facing functions on this class need a `Context`. Could you add `private val context: Context` to the constructor, and remove the Context parameter from the functions in the interface?

kevinmost commented on engine/src/main/java/com/google/android/fhir/sync/SyncJobImpl.kt:
```diff
@@ -0,0 +1,118 @@
+    return WorkManager.getInstance(context)
+      .getWorkInfosForUniqueWorkLiveData(uniqueWorkerName)
+      .asFlow()
+      .mapNotNull { if (it.isEmpty()) null else it[0] } // todo its always 0 ... would it be??
```
This can be

```kt
.mapNotNull { it.getOrNull(0) }
```

kevinmost commented on engine/src/main/java/com/google/android/fhir/sync/SyncJobImpl.kt:
```diff
@@ -0,0 +1,118 @@
+  }
+
+  private fun convertToState(workInfos: MutableList<WorkInfo>): State? {
+    for (wi in workInfos) {
+      if (wi.state != WorkInfo.State.ENQUEUED && wi.progress.keyValueMap.isNotEmpty()) {
+        val state = wi.progress.getString("StateType")!!
+        val stateData = wi.progress.getString("State")!!
+
+        val data: State = Gson().fromJson(stateData, Class.forName(state)) as State
+
+        return data
+      }
+    }
+    return null
```
This can be made functional:

```kt
return workInfos
  .asSequence()
  .filter { it.state != ENQUEUED && it.progress.keyValueMap.isNotEmpty() }
  .map { workInfo ->
    val state = workInfo.progress.getString("StateType")!!
    val stateData = workInfo.progress.getString("State")!!
    Gson().fromJson(stateData, Class.forName(state)) as State
  }
  .firstOrNull()
```

kevinmost commented on engine/src/test-common/java/com/google/android/fhir/resource/TestingUtils.kt:
```diff
@@ -63,4 +73,89 @@ class TestingUtils constructor(private val iParser: IParser) {
     return JSONArray(content)
   }
+
+  object testDataSourceImpl : DataSource {
```
Nit: `object` names should start with an uppercase letter

kevinmost commented on engine/src/test-common/java/com/google/android/fhir/resource/TestingUtils.kt:
```diff
@@ -63,4 +73,89 @@ class TestingUtils constructor(private val iParser: IParser) {
+    }
+  }
+
+  object testFhirEngineImpl : FhirEngine {
```
Same

kevinmost commented on engine/src/test-common/java/com/google/android/fhir/resource/TestingUtils.kt:
```diff
@@ -63,4 +73,89 @@ class TestingUtils constructor(private val iParser: IParser) {
+    override suspend fun saveLastSyncTimeStamp(lastSyncTimestamp: LocalDateTime) {}
+  }
+
+  object testCorruptDatasource : DataSource {
```
Same

kevinmost commented on engine/src/test/java/com/google/android/fhir/sync/SyncJobTest.kt:
```diff
@@ -0,0 +1,210 @@
+  private lateinit var workManager: WorkManager
+  private lateinit var fhirEngine: FhirEngine
+
+  @Mock private lateinit var database: Database
+  @Mock private lateinit var dataSource: DataSource
```
In Kotlin, prefer using the mockito-kotlin helpers

```kt
private val database = mock<Database>()
private val dataSource = mock<DataSource>()
```

I think you can also remove `MockitoAnnotations.openMocks(this)` on line 78

kevinmost commented on engine/src/test/java/com/google/android/fhir/sync/SyncJobTest.kt:
```diff
@@ -0,0 +1,210 @@
+    // Enqueued, Running [1 on start],
+    // Running [4 from emitted progress],
+    // Enqueued for successful runs
+    assertTrue(workInfoList.size == 7)
```
Please replace `org.junit.Assert.*` usages with `com.google.common.truth.Truth.assertThat` throughout this file.

Eg:

```kt
assertThat(workInfoList).hasSize(7)

assertThat(workInfoList[0].state).isEqualTo(WorkInfo.State.ENQUEUED)

assertThat(stateList[0]).isInstanceOf(State.Nothing::class.java)
```

maimoonak commented:
@kevinmost @yigit Thanks alot for the detailed and through feedback. It helped and made the code much more readable and precise. I have resolved all of the comments and this PR is ready for review. Aside from code and general design approach please review the approach being used in FhirSyncWorker and WorkManager usage in SyncJobImpl.

kevinmost commented on engine/src/main/java/com/google/android/fhir/DatastoreUtil.kt:
```diff
@@ -0,0 +1,49 @@
+  private val lastSyncTimestampKey by lazy { stringPreferencesKey("LAST_SYNC_TIMESTAMP") }
+
+  fun readLastSyncTimestamp(): OffsetDateTime? {
+    var millis: String?
+
+    millis = runBlocking { context.dataStore.data.first()[lastSyncTimestampKey] } ?: return null
```
This can be combined, and turned into a val

```kt
val millis = runBlocking { ... } ?: return null
```

kevinmost commented on engine/src/main/java/com/google/android/fhir/sync/FhirSyncWorker.kt:
```diff
@@ -30,8 +40,37 @@ abstract class FhirSyncWorker(appContext: Context, workerParams: WorkerParameter
+            workDataOf(
+              // send serialized state and type so that consumer can convert it back
+              "StateType" to it::class.java.name,
+              "State" to Gson().toJson(it)
```
Creating `Gson` instances can be expensive. Please move this up to a property on the class:

```kt
private val gson: Gson = Gson()
```

and then use that property here

```kt
"State" to gson.toJson(it)
```

kevinmost commented on engine/src/main/java/com/google/android/fhir/sync/FhirSyncWorker.kt:
```diff
@@ -30,8 +40,37 @@ abstract class FhirSyncWorker(appContext: Context, workerParams: WorkerParameter
+    val result = fhirSynchronizer.synchronize()
+
+    // await/join is needed to collect states completely
+    kotlin.runCatching { job.join() }.onFailure { Log.w(javaClass.name, it) }
```
Please pull out the tag for this log as well, so that it's not computed every time we log.

kevinmost commented on engine/src/main/java/com/google/android/fhir/sync/SyncJobImpl.kt:
```diff
@@ -0,0 +1,108 @@
+      ?.let {
+        val state = workInfo.progress.getString("StateType")!!
+        val stateData = workInfo.progress.getString("State")!!
+        Gson().fromJson(stateData, Class.forName(state)) as State
```
Similarly, please make a val for this `Gson` instance within this class so that it's only instantiated once for this class, instead of every time we call `convertToState`

kevinmost commented on engine/src/test/java/com/google/android/fhir/sync/SyncJobTest.kt:
```diff
@@ -0,0 +1,204 @@
+      .result
+      .get()
+
+    Thread.sleep(5000) // how to avoid it ???
```
Not sure, but what if you call

```kt
job1.join()
job2.join()
```

maimoonak replied:
@kevinmost I tried but it does not work

kevinmost commented on engine/src/test/java/com/google/android/fhir/sync/SyncJobTest.kt:
```diff
@@ -0,0 +1,204 @@
+
+    // State transition for failed job as below
+    // Nothing, Started, InProgress, Glitch, Failed (Error)
+    assertThat(res[0]).isInstanceOf(State.Nothing::class.java)
+    assertThat(res[1]).isInstanceOf(State.Started::class.java)
+    assertThat(res[2]).isInstanceOf(State.InProgress::class.java)
+    assertThat(res[3]).isInstanceOf(State.Glitch::class.java)
+    assertThat(res[4]).isInstanceOf(State.Failed::class.java)
```
You can also do

```kt
    assertThat(res.map { it::class.java })
      .containsExactly(
        State.Nothing::class.java,
        State.Started::class.java,
        State.InProgress::class.java,
        State.Glitch::class.java,
        State.Failed::class.java,
      ).inOrder()
```

Then you won't need the `assertThat(res).hasSize(5)` assertion below, either.

kevinmost reviewed (approved):
Once these comments are addressed, LGTM for merging.

Tarun-Bhardwaj commented:
@maimoonak , could you please advise a date by when the changes suggested by Kevin in his last review get incorporated?

maimoonak commented:
@kevinmost Please review. I have updated it and also added the Sync status to UI where for now it sends the updated state. 

<img src="https://user-images.githubusercontent.com/4829880/128135197-1e0fd882-2ab6-477a-a716-3533a032208f.jpg" data-canonical-src="https://user-images.githubusercontent.com/4829880/128135197-1e0fd882-2ab6-477a-a716-3533a032208f.jpg" width="200" height="400" />

kevinmost reviewed (approved):
Suggested some minor changes, but looks good once those are resolved.
